### PR TITLE
[LINALG] Add lowering for the aten.upsample_nearest2d op.

### PR DESCRIPTION
**Pull request by pashu123**

### CODE REVIEW

**Comment by ramiro050** [comment]
Can you resolve the conflicts so that the CI can run?

File: python/torch_mlir_e2e_test/test_suite/conv.py [hunk 1]
```diff
@@ -12,6 +12,7 @@
 
 
 class Conv2dNoPaddingModule(torch.nn.Module):
+
```
Comment (ramiro050) [inline python/torch_mlir_e2e_test/test_suite/conv.py:15]: Can you make all the style changes in this file in a separate PR? It makes it hard to know which changes are truly related to this PR

**Review by ramiro050 — CHANGES_REQUESTED** [review]

File: python/torch_mlir_e2e_test/test_suite/conv.py (hunk 1 above)
Comment (pashu123) [reply python/torch_mlir_e2e_test/test_suite/conv.py:15]: Sure.

Comment (pashu123) [reply python/torch_mlir_e2e_test/test_suite/conv.py]: Done.

File: lib/Conversion/TorchToLinalg/Linear.cpp [hunk 2]
```diff
@@ -152,6 +153,133 @@ class ConvertAtenFlipOp : public OpConversionPattern<AtenFlipOp> {
...
+    auto inputRank = input.getType().cast<RankedTensorType>().getRank();
+    Type elementType =
+        input.getType().cast<RankedTensorType>().getElementType();
+
+    SmallVector<Value> dims = getTensorSizes(rewriter, loc, input);
+    SmallVector<Value> scaleFactorsInt;
+
+    if (!adaptor.scale_factors().getType().isa<Torch::NoneType>()) {
+      SmallVector<double> scaleFactors;
+      if (!matchPattern(adaptor.scale_factors(),
```
Comment (ramiro050) [inline lib/Conversion/TorchToLinalg/Linear.cpp:221]: As far as I can tell, the lowering for this op does not need this to be a constant float list. It seems like all the floats are converted back to values. You should use `getListConstructElements` instead.

File: lib/Conversion/TorchToLinalg/Linear.cpp [hunk 3]
```diff
@@ -152,6 +153,133 @@ class ConvertAtenFlipOp : public OpConversionPattern<AtenFlipOp> {
 };
 } // namespace
 
+// Returns updated output dimension and scale_factor converted to integer.
+static std::pair<Value, Value> getScaledDims(OpBuilder &builder, Location loc,
```
Comment (ramiro050) [inline lib/Conversion/TorchToLinalg/Linear.cpp:157]: Getting the `scale_factor` converted to an int seems to be unrelated to the goal of this function, which is to scale the dims. I think it would be clearer to have this function simply return the scaled dim, and have the conversion to `int` of `scale_factor` happen in the lowering itself (since it's just one line of code)

File: python/torch_mlir_e2e_test/test_suite/conv.py [hunk 4]
```diff
@@ -705,3 +705,80 @@ def forward(self, inputVec, weight):
...
+class UpSampleNearest2dStaticSize(torch.nn.Module):
+
+    def __init__(self):
+        super().__init__()
+
+    @export
+    @annotate_args([
+        None,
+        ([1, 1, 4, 4], torch.float32, True),
```
Comment (ramiro050) [inline python/torch_mlir_e2e_test/test_suite/conv.py:717]: Is there a reason not to have all tests be fully dynamic?

File: python/torch_mlir_e2e_test/test_suite/conv.py [hunk 5]
```diff
@@ -705,3 +705,80 @@ def forward(self, inputVec, weight):
...
+    ])
+    def forward(self, inputVec):
+        return torch._C._nn.upsample_nearest2d(inputVec,
+                                               output_size=[11, 11],
+                                               scale_factors=None)
+
+
+@register_test_case(module_factory=lambda: UpSampleNearest2dStaticSize())
+def UpSampleNearest2dStaticSize_basic(module, tu: TestUtils):
+    module.forward(torch.randn(1, 1, 4, 4))
```
Comment (ramiro050) [inline python/torch_mlir_e2e_test/test_suite/conv.py:727]: use `tu.rand`. See:

https://github.com/llvm/torch-mlir/blob/main/docs/adding_an_e2e_test.md#test-case-function

**Review by ramiro050 — CHANGES_REQUESTED** [review]

File: include/torch-mlir/Dialect/Torch/IR/TorchOps.h [hunk 6]
```diff
@@ -150,6 +150,37 @@ m_TorchConstantIntList(SmallVectorImpl<int64_t> &bind_values) {
   return detail::torch_list_construct_op_binder(bind_values);
 }
 
+namespace detail {
+/// Matches the constant integers stored in a `torch.ListConstruct`.
```
[the remark this answers is not in the document]
Comment (pashu123) [reply include/torch-mlir/Dialect/Torch/IR/TorchOps.h]: Done.

File: lib/Conversion/TorchToLinalg/Linear.cpp [hunk 7]
```diff
@@ -152,6 +153,133 @@ class ConvertAtenFlipOp : public OpConversionPattern<AtenFlipOp> {
...
+                  ConversionPatternRewriter &rewriter) const override {
+
+    Location loc = op->getLoc();
+    Value input = adaptor.input();
+
+    Type resultType = getTypeConverter()->convertType(op.getResult().getType());
+
+    auto inputRank = input.getType().cast<RankedTensorType>().getRank();
+    Type elementType =
+        input.getType().cast<RankedTensorType>().getElementType();
```
[the remark this answers is not in the document]
Comment (pashu123) [reply lib/Conversion/TorchToLinalg/Linear.cpp]: Done.

File: lib/Conversion/TorchToLinalg/Linear.cpp [hunk 8]
```diff
@@ -152,6 +153,133 @@ class ConvertAtenFlipOp : public OpConversionPattern<AtenFlipOp> {
...
+
+    SmallVector<Value> dims = getTensorSizes(rewriter, loc, input);
+    SmallVector<Value> scaleFactorsInt;
+
+    if (!adaptor.scale_factors().getType().isa<Torch::NoneType>()) {
```
[the remark this answers is not in the document]
Comment (pashu123) [reply lib/Conversion/TorchToLinalg/Linear.cpp]: Done.

File: lib/Conversion/TorchToLinalg/Linear.cpp [hunk 9]
```diff
@@ -152,6 +153,133 @@ class ConvertAtenFlipOp : public OpConversionPattern<AtenFlipOp> {
...
+      SmallVector<double> scaleFactors;
+      if (!matchPattern(adaptor.scale_factors(),
+                        m_TorchConstantFloatList(scaleFactors)))
+        return rewriter.notifyMatchFailure(op,
+                                           "only constant dim lists supported");
+      for (unsigned i = 0; i < scaleFactors.size(); i++) {
+        auto scaledPair =
+            getScaledDims(rewriter, loc, dims[2 + i], scaleFactors[i]);
```
[the remark this answers is not in the document]
Comment (pashu123) [reply lib/Conversion/TorchToLinalg/Linear.cpp]: Done.

File: lib/Conversion/TorchToLinalg/Linear.cpp [hunk 10]
```diff
@@ -152,6 +153,133 @@ class ConvertAtenFlipOp : public OpConversionPattern<AtenFlipOp> {
...
+        scaleFactorsInt.push_back(scaledPair.first);
+        dims[2 + i] = scaledPair.second;
+      }
+    } else {
+
+      SmallVector<int64_t> outputDims;
+      if (!matchPattern(adaptor.output_size(),
```
[the remark this answers is not in the document]
Comment (pashu123) [reply lib/Conversion/TorchToLinalg/Linear.cpp]: I see, yes. Thanks for the suggestion.

File: python/torch_mlir_e2e_test/test_suite/conv.py (hunk 4 above)
Comment (pashu123) [reply python/torch_mlir_e2e_test/test_suite/conv.py:717]: I have added 2 static and 2 dynamic tests. If you want me to add all static I can do that.

Comment (ramiro050) [reply python/torch_mlir_e2e_test/test_suite/conv.py:717]: I would just keep the 2 dynamic ones and remove the static ones. The static case should just be a special case of the dynamic case, so if it works dynamically, it should work statically.

**Review by ramiro050 — APPROVED** [review]
Other than the removal of static tests and code to match float lists, this LGTM! Thanks!

File: include/torch-mlir/Dialect/Torch/IR/TorchOps.h [hunk 11]
```diff
@@ -150,6 +150,37 @@ m_TorchConstantIntList(SmallVectorImpl<int64_t> &bind_values) {
   return detail::torch_list_construct_op_binder(bind_values);
 }
 
+namespace detail {
+/// Matches the constant floats stored in a `torch.ListConstruct`.
+struct torch_list_construct_float_op_binder {
```
[the remark this answers is not in the document]
Comment (pashu123) [reply include/torch-mlir/Dialect/Torch/IR/TorchOps.h]: Makes sense. Done.

File: python/torch_mlir_e2e_test/test_suite/conv.py (hunk 4 above)
Comment (pashu123) [reply python/torch_mlir_e2e_test/test_suite/conv.py]: Done.

Comment (pashu123) [reply python/torch_mlir_e2e_test/test_suite/conv.py]: Converted the static ones to dynamic ones as they were covering different cases (same scaled_h, scaled_w vs different).

**Comment by pashu123** [comment]
@ramiro050 Any ideas why it's failing with the error message         ```
RuntimeError: Unimplemented Error: 
        	/main_checkout/torch-mlir/build/tools/torch-mlir/python/torch_mlir/csrc/base_lazy_backend/generated/shape_inference.cpp:79    std::vector<torch::lazy::Shape> torch::lazy::compute_shape_upsample_nearest2d(const at::Tensor &, at::OptionalIntArrayRef, c10::optional<at::ArrayRef<double>>)
```
